source/common/grpc/subscription_impl.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include "envoy/config/subscription.h"
+#include "envoy/event/dispatcher.h"
+
+#include "common/grpc/async_client_impl.h"
+
+#include "api/base.pb.h"
+
+namespace Envoy {
+namespace Grpc {
+
+template <class ResponseType, class ResourceType>
+class SubscriptionImpl : public Config::Subscription<ResponseType, ResourceType>,
+                         AsyncClientCallbacks<ResponseType> {
+public:
+  SubscriptionImpl(const envoy::api::v2::Node& node, Upstream::ClusterManager& cm,
+                   const std::string& remote_cluster_name, Event::Dispatcher& dispatcher,
+                   const google::protobuf::MethodDescriptor& service_method)
+      : SubscriptionImpl(node, new AsyncClientImpl<envoy::api::v2::DiscoveryRequest, ResponseType>(
```

Comment:
nit: for RAII reasons it would be better to pass a unique_ptr vs. raw pointer here and then move inside the other constructor definition.